<img src="http://oci02.img.iteso.mx/identidad_de_instancia_2018/ITESO/Logos%20ITESO/Logo-ITESO-Principal.jpg">

  ---

<center> <font color= #0000> <font size = 5> Ingeniería Financiera | ITESO </font> <br> <br>
   

<center> <font color= #000000> <font size = 4> Microestructura y Sistemas de Trading </font> <br> <br>

<center> <font color=#000000><font size=5> Laboratorio 2 <br> <br> Behavioral Finance en Trading </font> <br> <br>

<center> <font color=#000000><font size=4> I.F. Oscar E. Flores H. | if715029@iteso.mx </font> <br> <br>

  ---

# Abstract

En este trabajo se pretende hacer un análisis de comportamientos en entornos financieros. Se tomarán datos de transaccioones ejecutadas por algún algorítmo o trader (estos mismos proporcionados por Oanda y MetaTrader4), a partir de estos datos generamos un código que analice; los históricos que ha conseguido la cuenta, las medidas de atribuciión al desempeño (que tan bueno es el algorítmo/trader), y finanzas conductuales. 

# Análisis de histórico de cuenta

In [23]:
import funciones as fn

archivo = "archivo_tradeview_1.csv"
datos = fn.f_leer_archivo(archivo)
datos = fn.f_columnas_tiempos(datos)
datos = fn.f_columnas_pips(datos)
datos = fn.f_capital_acm(datos)
estadisticas = fn.f_estadisticas_ba(datos)
desempenio = fn.f_estadisticas_mad(datos)
sesgos = fn.f_sesgos_cognitivos(datos)

AttributeError: module 'funciones' has no attribute 'f_sesgos_cognitivos'

Desgargamos datos de traders/algoritmos y hacemos algunos calculos con ellos, entre los cálculos que hemos hecho se encuentra calcular la magnitud de pips según sea el instrumento que se está utilizando, la cantidad de ganancia/pérdida que se ha obtenido después de cerrar la operación, entre otras. A continuación se muestra la tabla de datos generada.

In [18]:
datos.head()

order            opentime type  size  symbol   openprice       s/l  \
0  24316034 2019-08-27 06:57:50  buy  0.04  xauusd  1529.21000  1523.000   
1  24334215 2019-08-27 17:06:20  buy  0.01  eurjpy   117.66400   116.950   
2  24342155 2019-08-27 19:25:39  buy  0.05  eurjpy   117.34600   116.950   
3  24373260 2019-08-28 17:11:33  buy  0.01  eurusd     1.10791     1.104   
4  24392415 2019-08-29 06:35:27  buy  0.01  gbpusd     1.22024     1.208   

         t/p           closetime  closeprice  commission  taxes  swap  profit  \
0  1543.0000 2019-08-27 09:16:01  1533.69000         0.0      0     0   17.92   
1   118.7000 2019-08-29 06:38:16   117.37400         0.0      0     0   -2.73   
2   119.7000 2019-08-28 21:01:24   117.45200         0.0      0     0    5.00   
3     1.1175 2019-08-29 06:38:28     1.10818         0.0      0     0    0.27   
4     1.2370 2019-08-29 06:37:23     1.21990         0.0      0     0   -0.34   

     tiempo  pips  pips_acm  profit_acm  capital_acm  
0    8291.0  44.8      44.8       17.92      5017.92  
1  135116.0 -29.0      15.8       15.19      5015.19  
2   92145.0  10.6      26.4       20.19      5020.19  
3   48415.0   2.7      29.1       20.46      5020.46  
4     116.0  -3.4      25.7       20.12      5020.12

### estadísticas básicas de la tabla generada:

In [19]:
estadisticas['df1']

Ops totales              Ganadoras  \
Valor                         84                     69   
Descripcion  Operaciones totales  Operaciones ganadoras   

                                 Ganadoras_c                     Ganadoras_s  \
Valor                                     24                              45   
Descripcion  Operaciones ganadoras de compra  Operaciones ganadoras de venta   

                         Perdedoras                      Perdedoras_c  \
Valor                            15                                15   
Descripcion  Operaciones perdedoras  Operaciones perdedoras de compra   

                                Perdedoras_s  \
Valor                                      0   
Descripcion  Operaciones perdedoras de venta   

                                         Mediana_profit  \
Valor                                             1.205   
Descripcion  Mediana de rendimeintos de las operaciones   

                                   Mediana_pips  \
Valor                                   1252.38   
Descripcion  Mediana de pips de las operaciones   

                                     r_efectividad  \
Valor                                     0.821429   
Descripcion  Ganadoras Totales/Operaciones Totales   

                                      r_proporcion  \
Valor                                          4.6   
Descripcion  Ganadoras Totales/ Perdedoras Totales   

                                    r_efectividad_c  \
Valor                                      0.615385   
Descripcion  Ganadoras Compras/ Operaciones Totales   

                                   r_efectividad_v  
Valor                                            1  
Descripcion  Ganadoras Ventas/ Operaciones Totales

En este caso podemos observar que son pocas operaciones las que han sido ejecutadas (únicamente 84), aunque realmente la precisión de las operaciones es muy buena, ya que tiene 69 ganadoras (esto equivale al 82.14% de efectividad). Además de esto podemos notar que es prácticamente indiferente entre si ejecutar posiciones de compra o ejecutar posiciones de venta. Se puede notar también que no se tiene  ninguna posición de venta con rendimientos negativos, sin embargo, las posiciones de compra con rendimientos negativos equivalen al 17% de las operaciones totales.

Podemos observar tambien que la mediana de profit es pequeña, aunque positiva. El solo hecho de que la mediana sea positiva nos dice que el algorítmo ha tenido un buen desempeño. Las posiciones en las que ha tenido rendimientos negativos no fueron tan perjudiciales como fueron las posiciones con rendimientos positivos. 

### Rankings:

In [20]:
estadisticas['df2']

btcusd  eurgbp  usdmxn    xauusd    gbpusd  usdjpy    eurusd    audusd  \
rank     1.0     1.0     0.8  0.736842  0.666667     0.6  0.478261  0.333333   

        eurjpy    usdcad  gbpjpy  
rank  0.285714  0.285714    0.25

En la tabla anterior se pretende comparar la tasa de efectividad de cada uno de los instrumentos utilizados. Se puede notar que existen dos instrumentos que tienen 'muy buena tasa' de efectividad, sin embargo, esta tasa se ve sesgada debido a que las muestras son muy reducidas (información que no está contenida explícitamente en la tabla mostrada). Mismo caso con los que aparentemente tienen 'muy mal desempeño' o los instrumentos de extrema derecha.  

# Medidas de atribución al desempeño

Para este apratado se generó una tabla con las medidas de atribución al desempeño más conocidas como son; sharpe ratio, sortino ratios (positivos y negativos), y drawdown, entre otros. 

In [21]:
desempenio

sharpe   sortino_c   sortino_s  drawdown  drawup  drawdown_pips  \
Valor -102.133745 -151.913054 -124.554641    -111.3  254.13        -1174.2   

       drawup_pips  
Valor      1757.33

In [22]:
256/5000

0.0512

Se puede notar que el sharpe en este caso es negativo. Esto se debe a que el supuesto para el calculo de sharpe es un supuesto "incorrecto" debido a que la medida de comparación debe considerar los rendimientos generados por transacción como si estos fueran generados diariamente para ser multiplicado por 30 (los días del mes) y compararlo con la tasa de referencia proporcionada por el docente (8%). Cabe resaltar que los rendimientos son por transacción y no por día, hace falta generar una columna con la información del rendimiento diario para poder hacerla comparable. En cuento a drawdown, podemos notar que es únicamente de -113 que equivale a un 2.26% de bajas totales. El drawup por su parte equivale al 5.12%.  En el mejor de los casos tuvimos una 'plusvalía' de 5%. De la misma forma se puede interpretar los drawdown en pips y drawup


# Behavioral Finance

En este trabajo se pretende encontrar sesgos cognitivos. 
Como propuesta se hace una matriz de probabilidades de Markov. En las filas tenemos la situación "actual" y en las columnas tenemos la probabilidad de seguir en la misma situación o de cambiar a otra situación.

Podemos asumir que sí existe un sesgo cognitivo, esto se observa en la probabilidad empírica; perder dinero dado que la última transacción realizada tuvo perdidas tiene una probabilidad del 57.89%, del mismo modo, dada la situación de haber ganado dinero en la última transacción realizada, la probabilidad de ganar dinero en la siguiente transacción será de 64.44%. 

Se puede entonces notar que sí hay un patrón con un poco de sesgo, cuando se gana dinero es muy probable que sigan en rachas de transacciones con rendimientos positivos. Del mismo modo, una vez que entra en 'racha' de perdidas, se mantendrá en ella aunque con menor frecuencia que cuando se tienen rachas de rendimientos positivos. 